docs/proposed/psa-driver-interface.md

```diff
@@ -297,25 +297,251 @@ TODO
+    psa_key_derivation_step_t step,
+    uint64_t *value);
+```
+
+These functions take the following parameters:
+
+* The first parameter `inputs` must be a pointer passed by the core to a key derivation driver setup entry points which has not returned yet.
+* The `step` parameter indicates the input step whose content the driver wants to retrieve. The type of the input step must be compatible with the function:
+    * `psa_crypto_driver_key_derivation_get_input_integer` for integer inputs (steps that the application passes with `psa_key_derivation_input_integer()`).
+    * `psa_crypto_driver_key_derivation_get_input_size` and `psa_crypto_driver_key_derivation_get_input_bytes` for data inputs (steps that the application passes with `psa_key_derivation_input_bytes()` or `psa_key_derivation_input_key()`, excluding key inputs from the same secure element).
+    * `psa_crypto_driver_key_derivation_get_input_key` for key inputs (steps that the application passes with `psa_key_derivation_input_key()`, only for secure element drivers receiving a key from the same secure element).
+* On a successful invocation of `psa_crypto_driver_key_derivation_get_input_size`, the core sets `*size` to the size of the desired input in bytes.
+* On a successful invocation of `psa_crypto_driver_key_derivation_get_input_bytes`, the core fills the first *N* bytes of `buffer` with the desired input and sets `*buffer_length` to *N*, where *N* is the length of the input in bytes. The value of `buffer_size` must be at least *N*, otherwise this function fails with the status `PSA_ERROR_BUFFER_TOO_SMALL`.
+* On a successful invocation of `psa_crypto_driver_key_derivation_get_input_key`, the core sets `*key_buffer` to a pointer to a buffer containing the key context and `*key_buffer_size` to the size of the key context in bytes. The key context buffer remains valid for the duration of the driver entry point. If the driver needs to access the key context after the current entry point returns, it must make a copy of the key context.
+* On a successful invocation of `psa_crypto_driver_key_derivation_get_input_integer`, the core sets `*value` to the value of the desired input.
+
+These functions can return the following statuses:
+
+* `PSA_SUCCESS`: the call succeeded and the desired value has been copied to the output parameter (`size`, `buffer`, `value` or `p_key_buffer`) and if applicable the size of the value has been writen to the applicable parameter (`buffer_length`, `key_buffer_size`).
+* `PSA_ERROR_INSUFFICIENT_DATA`: the driver called `psa_crypto_driver_key_derivation_get_input_key` on a data input step which is available as a bytes input, or the driver called ``psa_crypto_driver_key_derivation_get_input_size` or `psa_crypto_driver_key_derivation_get_input_bytes` on a data input step which is available as a key input. This is not a fatal error and the driver is expected to call the appropriate function(s) instead.
```

Comment:
I agree that this is not the intended purpose of `PSA_ERROR_INSUFFICIENT_DATA`, which is when a caller asks for a certain _amount_ of data but only a smaller amount is available. Here, the caller is asking for an entry in a (conceptual) table, and this entry does not exist. So it would be more natural to use `PSA_ERROR_DOES_NOT_EXIST`. However, here, this is already taken for a different error. Furthermore, it isn't a good fit, since this error case also indicates that while there is no entry for exactly what the driver is asking, there is one for something closely related.

There are three cases where the caller might ask for an entry that doesn't exist, and I think it's useful to distinguish the three. Let's look at what a driver might do if it receives this error. I distinguish two typical kinds of drivers: an algorithm-aware driver that's actually doing the cryptographic operation, and a pass-through driver that's just forwarding stuff it doesn't understand to a companion system that will perform the cryptographic operation.

* The driver asked for something that doesn't make sense, such as an INFO input for an algorithm that doesn't take an INFO input, or an integer value for a step that takes a byte-string input. Currently `INVALID_ARGUMENT`.
    * Typical reaction for an algorithm-aware driver: it did something wrong and should quit with a fatal error.
    * Typical reaction for a pass-through driver: move on to the next thing.
* The driver asked for something that's optional, and it happens that the application didn't pass that optional input. Currently `DOES_NOT_EXIST`.
    * Typical reaction for an algorithm-aware driver: either remember the input as missing or fill in with a default value.
    * Typical reaction for a pass-through driver: either move on to the next thing or indicate the value as missing.
* The driver asked for something that's present, but in a different form from the way the driver is querying it. Currently `INSUFFICIENT_DATA`.
    * Typical reaction for a driver (either kind): query the data using the correct function.

I don't remember why I chose `INSUFFICIENT_DATA`. Maybe because it vaguely fits the framework, in that there is some data there but not exactly what is wanted? I guess `INVALID_ARGUMENT` would work better, but it's really a better fit where it's used. I can't think of a less bad fit and I'd rather not  invent a new error code just for this.

A possible solution would be to add one more function:
```
enum psa_crypto_driver_key_derivation_input_type_t {
    PSA_KEY_DERIVATION_INPUT_TYPE_INVALID,
    PSA_KEY_DERIVATION_INPUT_TYPE_OMITTED,
    PSA_KEY_DERIVATION_INPUT_TYPE_BYTES,
    PSA_KEY_DERIVATION_INPUT_TYPE_KEY,
    PSA_KEY_DERIVATION_INPUT_TYPE_INTEGER,
};
psa_crypto_driver_key_derivation_input_type_t psa_crypto_driver_key_derivation_get_input_type(
    const psa_crypto_driver_key_derivation_inputs_t *inputs,
    psa_key_derivation_step_t step);
```
`psa_crypto_driver_key_derivation_get_input_type` would return the actual type of the input, or `INVALID` or `OMITTED` if applicable. On the one hand, it's one more type and one more function. On the other hand, it means that the get-data functions can just return a “you did something wrong” error to the caller, without having to convey what was wrong exactly.

Reply:
I think the other two error codes are correct, which then does IMO lead to the conclusion that `INSUFFICIENT_DATA` is the best type for the third case.

I do quite like the option of `psa_crypto_driver_key_derivation_get_input_type`, since it eliminates the trial-and-error the driver needs to do to work out what type a particular input is, though I agree that adding another type for a single function is somewhat troubling.

I think both options are acceptable, and don't have a strong preference.